File: Code/Prototype/assetallocationtask.py
################################################################################
# Description: Asset allocation task, built on the PyBrain architecture.
# Author:      Pierpaolo Necchi
# Email:       pierpaolo.necchi@gmail.com
# Date:        dom 15 mag 2016 14:44:13 CEST
################################################################################

import numpy as np
import pandas as pd

from pybrain.rl.environments.episodic import EpisodicTask
from tradingperformance import portfolioSimpleReturn


class AssetAllocationTask(EpisodicTask):
    """ The asset allocation task on a financial market. The agent can trade I+1
        different assets only at discrete times by specifying a portfolio
        allocation. The state that the agent observes is given by the past
        returns of the tradable assets augmented with the current allocation.
        The reward function is the portfolio log-return."""

    def __init__(self,
                 environment,
                 deltaP,
                 deltaF,
                 deltaS,
                 discount,
                 backtest=False):
        """ Standard constructor for the asset allocation task.

        Args:
            environment (Environment): market environment object
            deltaP (double): proportional transaction costs rate
            deltaF (double): fixed transaction cost rate
            deltaS (double): short selling borrowing cost rate
            discount (double): discount factor
            backtest (bool): flag for training mode or test mode
        """
        # Initialize episodic task
        EpisodicTask.__init__(self, environment)

        # Transaction costs
        self.deltaP = deltaP
        self.deltaF = deltaF
        self.deltaS = deltaS

        # Discount factor
        self.discount = discount

        # Backtesting
        self.backtest = backtest

        # Report stores allocations and portfolio log-returns for backtesting
        self.report = pd.DataFrame(columns=list(self.env.data.columns) +
                                           ['ptfLogReturn'])

        # Initialize allocation
        self.initializeAllocation()

    def getObservation(self):
        """ An augmented observation of the underlying environment state that
            also includes the current portfolio weights, right before
            realloacation.

        Returns:
            state (np.array): the augmented state (size (P+1) * (I+1))
        """
        # Observe past asset returns from the environment
        pastReturns = EpisodicTask.getObservation(self)

        # Return augmented state
        return np.concatenate((pastReturns, self.currentAllocation))

    def performAction(self, action):
        """ Perform action on the underlying environment, i.e specify new asset
        allocation.

        Args:
            action (np.array): new allocation
        """
        # Cache new asset allocation for computing rewards
        self.newAllocation = action
        # Perform action
        EpisodicTask.performAction(self, action)

    def getReward(self):
        """ Function that returns the portfolio simple returns associated with
            the specified allocation.

        Returns:
            ptfSimpleReturn (double): portfolio simple return
        """
        # Retrieve current time step asset returns
        assetReturns = self.env.getAssetReturns()

        # Compute portfolio return associated to new allocation
        ptfSimpleReturn = portfolioSimpleReturn(assetReturns,
                                                self.currentAllocation,
                                                self.newAllocation,
                                                self.deltaP,
                                                self.deltaF,
                                                self.deltaS)
        # Update allocation weights
        self.currentAllocation = self.newAllocation * \
            (1.0 + assetReturns) / (1.0 + ptfSimpleReturn)

        # Compute portfolio log-return
        ptfLogReturn = np.log(1.0 + ptfSimpleReturn)

        # Store allocation and return when backtesting
        if self.backtest:
            reportEntry = np.append(self.newAllocation, ptfLogReturn)
            currentDate = self.env.getDate()
            self.report.ix[currentDate, :] = reportEntry

        return ptfLogReturn


    def isFinished(self):
        """ Function that checks if the current episode is over. To define an
            episode, we consider a receding horizon of length T.

        Returns:
            over (bool): flag that indicates if the current episode is over
        """
        if self.env.currentTimeStep >= self.env.finalTimeStep:
            self.initializeAllocation()
            return True
        else:
            return False

    def initializeAllocation(self):
        """ Initialize portfolio allocation at the beginning of an episode
        """
        # Equally weighted portfolio
        # self.currentAllocation = np.ones(self.env.indim) / self.env.indim
        # Everything invested in the risk-free asset
        self.currentAllocation = np.zeros(self.env.indim)
        self.currentAllocation[0] = 1.0
        # Random initialization
        # temp = np.random.rand(self.env.indim)
        # self.currentAllocation = temp / np.sum(temp)

    def setEvaluationInterval(self, start, end):
        """ Set the time interval to be considered in the evaluation. This
        function is used to change the evaluation interval during the backtest
        procedure.

        Args:
            start (int): start time index
            end (int): end time index
        """
        self.env.setEvaluationInterval(start, end)

    def trainingMode(self):
        """ Set the task in training mode
        """
        self.backtest = False

    def backtestMode(self):
        """ Set the task in backtest mode
        """
        self.backtest = True

File: Code/Thesis/src/assetallocationtask.cpp
#include <thesis/assetallocationtask.h>
#include <armadillo>
#include <math.h>      /* log */
#include <limits>      /* numeric_limits */

void AssetAllocationTask::initializeStatesCache()
{
	// Initialize past market states
	arma::vec proxyAction(dimAction);
	arma::vec tempState(dimState);
	for(size_t i = 0; i < numDaysObserved; ++i)
	{
        // Get market state
		market.getState(tempState);
        pastStates.rows(i * dimState, (i + 1) * dimState - 1) = tempState;

		// Move to the next time step
		market.performAction(proxyAction);
	}

	// Initialize current market state
    market.getState(currentState);
}

void AssetAllocationTask::initializeAllocationCache()
{
	currentAllocation.zeros();
	currentAllocation(0) = 1.0;
}

AssetAllocationTask::AssetAllocationTask (MarketEnvironment const & market_,
										  double deltaP_,
										  double deltaF_,
										  double deltaS_,
										  size_t numDaysObserved_)
	: market(market_),
	  deltaP(deltaP_),
	  deltaF(deltaF_),
	  deltaS(deltaS_),
	  numDaysObserved(numDaysObserved_)
{
	// Dimensions of observation and action spaces
	dimState = market.getDimState();
	dimAction = market.getDimAction();
	dimPastStates = numDaysObserved * dimState;
	dimObservation = dimPastStates + dimState + dimAction;

	// Initialize state cache variables
	pastStates.set_size(dimPastStates);
	currentState.set_size(dimState);
	initializeStatesCache();

	// Initialize allocation cache variables
	currentAllocation.set_size(dimAction);
	newAllocation.set_size(dimAction);
	initializeAllocationCache();
}

void AssetAllocationTask::getObservation (arma::vec &observation) const
{
	// Past states
	observation.rows(0, dimPastStates-1) = pastStates;

	// Current state
	observation.rows(dimPastStates, dimPastStates + dimState - 1) = currentState;

	// Current allocation
	observation.rows(dimPastStates + dimState, observation.size() - 1) = currentAllocation;
}

void AssetAllocationTask::performAction (arma::vec const &action)
{
	// Cache new allocation
	newAllocation = action;

	// Broadcast action to underlying environment
	market.performAction(action);
}

double AssetAllocationTask::getReward ()
{
	// Update past states with current state
	pastStates.rows(0, (numDaysObserved - 1) * dimState - 1) =
		pastStates.rows(dimState, pastStates.size() - 1);
	pastStates.rows((numDaysObserved - 1) * dimState, pastStates.size() - 1) =
		currentState;

	// Observe new market state
	market.getState(currentState);

	// Compute portfolio simple return
	double portfolioSimpleReturn = computePortfolioSimpleReturn();

	// Update allocation weights
	currentAllocation = newAllocation % (1.0 + currentState) /
						(1.0 + portfolioSimpleReturn);

	// Return portfolio log-return
	return log(1.0 + portfolioSimpleReturn);
}

void AssetAllocationTask::setEvaluationInterval (size_t startDate_,
												 size_t endDate_)
{
	market.setEvaluationInterval(startDate_, endDate_);
	initializeStatesCache();
	initializeAllocationCache();
}

double AssetAllocationTask::computePortfolioSimpleReturn () const
{
	// Proportional transaction costs
	double proportionTransactionCosts = deltaP *
		arma::sum(arma::abs(newAllocation - currentAllocation));

	// Fixed transaction costs
	double fixedTransactionCosts = 	deltaF *
		(!arma::approx_equal(currentAllocation, newAllocation, "absdiff",
							 std::numeric_limits<double>::epsilon()));

	// Short-selling fees
	double shortPositionsWeight = 0.0;
	for(size_t i = 0; i < newAllocation.size(); ++i)
		if (newAllocation(i) < 0.0)
			shortPositionsWeight += - newAllocation(i);
	double shortTransactionCosts = deltaS * shortPositionsWeight;

	// Trading profit & loss
	double tradingPL = arma::dot(newAllocation, currentState);

	// Compute simple portfolio return
	double portfolioSimpleReturn = tradingPL
								 - proportionTransactionCosts
								 - fixedTransactionCosts
								 - shortTransactionCosts;
	return portfolioSimpleReturn;
}
